File: LeetCode/python3/652.py
class Solution:
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """

        if not root:
            return []
        tree_list = []
        cnt = collections.Counter()

        def Traverse(node):
            if not node:
                return '#'
            serialized_tree = "{}{}{}".format(
                node.val, Traverse(node.left), Traverse(node.right))
            cnt[serialized_tree] += 1
            if cnt[serialized_tree] == 2:
                tree_list.append(node)
            return serialized_tree

        Traverse(root)
        return tree_list

File: LeetCode/cpp/652.cpp
class Solution {
public:
    vector<TreeNode *> sub;
    unordered_map<string, int> m;

    vector<TreeNode *> findDuplicateSubtrees(TreeNode *root) {
        sub.clear();
        m.clear();
        Traverse(root);
        return sub;
    }

    string Traverse(TreeNode *node) {
        if (!node)
            return "#";
        string s;
        s += to_string(node->val);
        string left = Traverse(node->left);
        string right = Traverse(node->right);
        s += ',' + left + ',' + right;

        if (m[s] == 1)
            sub.push_back(node);
        ++m[s];
        return s;
    }
};
